### PR TITLE
[PA-751] Add N number of labels to the namespace, perform backup and restore

### PR DESCRIPTION
**Pull request by apimpalgaonkar**
**What this PR does / why we need it**:

* Add 1000 labels to namespace in batch of 10 
* Create backup and restore

**Which issue(s) this PR fixes** (optional)
Closes #
https://portworx.atlassian.net/browse/PA-751

Jenkins - 
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/937/consoleFull

### CODE REVIEW

File: tests/backup/backup_helper.go [hunk 1]
```diff
@@ -1445,3 +1445,25 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
 	_, err := task.DoRetryWithTimeout(backupDeletionSuccessCheck, backupDeleteTimeout, backupDeleteRetryTime)
 	return err
 }
+func AddMultipleLabelsToNS(number int, namespace string) (int, error) {
```
Comment (kshithijiyer-px) [inline tests/backup/backup_helper.go:1448]: Add comment explaining what this func does.

File: tests/backup/backup_resiliency_test.go [hunk 2]
```diff
@@ -1045,3 +1045,72 @@ var _ = Describe("{ScaleMongoDBWhileBackupAndRestore}", func() {
 		CleanupCloudSettingsAndClusters(backupLocationMap, cloudCredName, cloudCredUID, ctx)
 	})
 })
+
+// AddMultipleNamespaceLabel adds labels to namespace in range of 1000
```
Comment (kshithijiyer-px) [inline tests/backup/backup_resiliency_test.go:1049]: Incorrect comment please check.

File: tests/backup/backup_resiliency_test.go [hunk 3]
```diff
@@ -1045,3 +1045,72 @@ var _ = Describe("{ScaleMongoDBWhileBackupAndRestore}", func() {
...
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Adding [%v] labels to namespaces [%v]", addedLabels, batchNs))
+			}
+		})
+		Step("Verifying number of labels added to namespace", func() {
+			log.InfoD("Verifying number of labels added to namespace")
+			for _, namespace := range bkpNamespaces {
+				labelMap, err := Inst().S.GetNamespaceLabel(namespace)
+				length := len(labelMap)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching labels %v for namespace %v", labelMap, namespace))
+				dash.VerifyFatal(length-3, desiredNumLabels, fmt.Sprintf("Verifying number of added labels to desired labels for namespace %v", namespace))
```
Comment (kshithijiyer-px) [inline tests/backup/backup_resiliency_test.go:1104]: I still don't understand how is backup being used here.

**Review by kshithijiyer-px — CHANGES_REQUESTED** [review]

File: tests/backup/backup_restore_basic_test.go [hunk 4]
```diff
@@ -1318,3 +1318,128 @@ var _ = Describe("{BackupSyncBasicTest}", func() {
...
+		})
+		Step("Taking a backup of application", func() {
+			log.InfoD("Taking a backup of application")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Unable to fetch px-central-admin ctx")
+			namespaceLabel := MapToKeyValueString(labelMap)
+			backupName = fmt.Sprintf("%s-%v", "backup", time.Now().Unix())
+			err = CreateBackupWithNamespaceLabel(backupName, SourceClusterName, backupLocationName, backupLocationUID,
+				nil, orgID, clusterUid, "", "", "", "", namespaceLabel, ctx)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying backup [%s] creation with label [%s]", backupName, namespaceLabel))
```
Comment (vpinisetti-px) [inline tests/backup/backup_restore_basic_test.go:1421]: I can see only few labels from this output. Is this latest one ?
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/765/consoleFull

```
20:26:48  INFO[2023-03-29 14:56:47] Verifying : Description : Verifying backup [backup-1680101776] creation with label [3-8af10fa7-5427-4a9e-b708-8fc53f3440b3=c38c2986-c3d6-4a42-85e7-6705ebde6161,4-5fe729ea-15dd-4e98-9841-76cf0c4258c8=2101e216-f072-42f9-aab7-2de4091bf4c5,8-33a63c91-6067-41c2-a1f7-bcdb5155fbec=11303821-8675-4605-86fb-22d0d2a2a9e9,9-887fcc2e-7b39-4bef-8256-ed8d846a6c40=6adec6e2-1213-48c4-aa08-82505183c2bc,0-2c2abbd4-ab34-4e14-864c-898f5c426904=390a566d-7598-491c-8f30-12b4c990dda6,1-36df360c-08b3-4544-ad2e-653a2c80de48=876a942a-d366-43bf-963b-741212ba4fe0,2-ff9ee144-8569-4fd4-97d7-d657defbe2d3=1d646a15-65b2-4232-80e7-693a8615f458,5-1a7568da-b760-4fa5-bcaa-d787318704f5=ce14d502-5a53-42a3-aad4-3e20d46fb5a8,6-f4da76a7-2c7a-44ed-b261-b6de15d4775f=19f18e44-4cfe-4fcf-9bba-672081101e96,7-7075e507-2431-41db-bdc7-1d67f637cdec=e5b5d1dc-4660-4ba2-b49d-48b9df0b07c7] 
```

File: drivers/scheduler/k8s/k8s.go [hunk 5]
```diff
@@ -6492,13 +6492,22 @@ func getLabelsFromNodeAffinity(nodeAffSpec *v1.NodeAffinity) map[string]string {
 	return label
 }
 
+// MergeMaps merges two maps
+func MergeMaps(m1 map[string]string, m2 map[string]string) map[string]string {
```
Comment (vpinisetti-px) [inline drivers/scheduler/k8s/k8s.go:6496]: Rebase and cleanup these helper methods as all these are part of different PR

**Review by vpinisetti-px — CHANGES_REQUESTED** [review]

File: tests/backup/backup_restore_basic_test.go [hunk 6]
```diff
@@ -1442,3 +1442,131 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching [%s] cluster uid", appClusterName))
+			log.InfoD("Uid of [%s] cluster is %s", appClusterName, clusterUid)
+		})
+		Step("Taking a backup of application", func() {
+			log.InfoD("Taking a backup of application")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Unable to fetch px-central-admin ctx")
+			namespaceLabel := MapToKeyValueString(fetchedLabelMap)
+			backupName = fmt.Sprintf("%s-%v", "backup", time.Now().Unix())
+			err = CreateBackupWithNamespaceLabel(backupName, SourceClusterName, backupLocationName, backupLocationUID,
```
Comment (vpinisetti-px) [inline tests/backup/backup_restore_basic_test.go:1546]: Better try one more backup with schedule and do default & custom restore with the same backup.

**Review by vpinisetti-px — CHANGES_REQUESTED** [review]

**Review by vpinisetti-px — APPROVED** [review]

**Review by skonda-px — APPROVED** [review]
LGTM

File: tests/backup/backup_helper.go [hunk 7]
```diff
@@ -2027,3 +2027,23 @@ func MapToKeyValueString(m map[string]string) string {
 	}
 	return strings.Join(pairs, ",")
 }
+
+// FetchNamespaceFromBackup fetches the namespace from backup
+func FetchNamespaceFromBackup(ctx context.Context, backupName string, orgID string) ([]string, error) {
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:2032]: nit: It should be `FetchNamespacesFromBackup `

File: tests/backup/backup_restore_basic_test.go [hunk 8]
```diff
@@ -1442,3 +1442,188 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+	})
+	JustAfterEach(func() {
+		defer EndPxBackupTorpedoTest(contexts)
+		ctx, err := backup.GetAdminCtxFromSecret()
+		log.FailOnError(err, "Unable to px-central-admin ctx")
+		scheduleUid, err := GetScheduleUID(scheduleName, orgID, ctx)
+		log.FailOnError(err, "Error while getting schedule uid %s", scheduleName)
+		err = DeleteSchedule(scheduleName, scheduleUid, orgID)
+		dash.VerifySafely(err, nil, fmt.Sprintf("Verification of deleting backup schedule - %s", scheduleName))
+		time.Sleep(1 * time.Minute)
```
Comment (mkoppal-px) [inline tests/backup/backup_restore_basic_test.go:1617]: Why is there `time.Sleep` here?

File: tests/backup/backup_restore_basic_test.go [hunk 9]
```diff
@@ -1442,3 +1442,188 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+			}
+		}
+	})
+	It("Add multiple labels to namespaces, perform manual backup, schedule backup using namespace label and restore", func() {
+		providers := getProviders()
+		Step("Validate applications", func() {
+			log.InfoD("Validate applications")
+			ValidateApplications(contexts)
+		})
+		Step("Adding labels to namespaces in multiple of 10 until 1000", func() {
```
Comment (mkoppal-px) [inline tests/backup/backup_restore_basic_test.go:1499]: Why do we have to add in batches?

Comment (apimpalgaonkar) [reply tests/backup/backup_restore_basic_test.go:1499]: Test case says add in batch of 10 till 1000

File: tests/backup/backup_restore_basic_test.go (hunk 8 above)
Comment (apimpalgaonkar) [reply tests/backup/backup_restore_basic_test.go:1617]: Removed, added that for debugging purposes.

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Review by kshithijiyer-px — APPROVED** [review]
LGTM

**Pull request merged**